Add label to SFN map state

u-minor opened the pull request:
<!-- Please refer to the contribution guidelines before raising a PR: https://github.com/localstack/localstack/blob/master/docs/CONTRIBUTING.md -->

<!-- Why am I raising this PR? Add context such as related issues, PRs, or documentation. -->
## Motivation

Currently the SFN interpreter lacks support for Label field in distributed Map state.
This PR introduces support for this.

Closes: #11029

<!-- What changes does this PR make? How does LocalStack behave differently now? -->
## Changes

- Added `Label` to the ASL lexser and parser.
- Added relevant test.

MEPalma commented on tests/aws/services/stepfunctions/templates/scenarios/statemachines/map_state_label.json5:
```diff
@@ -0,0 +1,34 @@
+{
```
Great to see snapshot tests for this new feature.
We can get away with simpler/faster tests that only use `Pass` states, such as for `map_state_config_distributed_parameters.json5`.

u-minor replied:
Fixed in [c1129ef](https://github.com/localstack/localstack/pull/11073/commits/c1129efb638c6b82d40296a996560dd681a46ab0).

MEPalma commented on tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py:
```diff
@@ -1121,6 +1121,35 @@ def test_map_state_tolerated_failure_percentage_path(
             exec_input,
         )
 
+    @markers.aws.validated
```
Great, let's simplify this test case by avoiding s3 setups and have the map run execute simple `pass` states.

u-minor replied:
Fixed in [c1129ef](https://github.com/localstack/localstack/pull/11073/commits/c1129efb638c6b82d40296a996560dd681a46ab0).

MEPalma commented on tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py:
```diff
@@ -1121,6 +1121,35 @@ def test_map_state_tolerated_failure_percentage_path(
         )
 
+    @markers.aws.validated
+    def test_map_state_label(
```
Would be good to add in `aws.services.stepfunctions.v2.test_sfn_api_map_run.TestSnfApiMapRun` some api validation for state machine creation that have invalid `Label` fields (empty, with illegal characters, too long, incorrect type). The tests could check for raising an exception on creation and skip the validation of the error message as we do not have a static analyser yet. (see comment above about assertions on Label instantiation)

u-minor replied:
Fixed in [df3f18d](https://github.com/localstack/localstack/pull/11073/commits/df3f18ddb008ee9572898155f82643f608046ef7).

MEPalma reviewed (commented):
Thank you for working on adding support for `Label` declarations in StepFunctions!

I've made a few change requests. First, let's standardise the `Label` type to be nullable when unset, which aligns with the provider's usual pattern. Currently, unset labels default to an empty string. Second, please add a label validity check and a parameterised test to ensure basic conditions are met. Lastly, simplify the test case by removing the need to set up S3 resources, as this is unnecessary for validating this feature.

I look forward to seeing these changes and having StepFunctions support `Label` declarations!

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/iteration/distributed_iteration_component.py:
```diff
@@ -76,6 +77,7 @@ def __init__(
         self.item_reader = item_reader
         self.tolerated_failure_count = tolerated_failure_count
         self.tolerated_failure_percentage = tolerated_failure_percentage
+        self.label = label
```
@MEPalma 
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/iteration/itemprocessor/map_run_record.py:
```diff
@@ -129,13 +130,13 @@ def __init__(
 
     @staticmethod
-    def _generate_map_run_arns(state_machine_arn: Arn) -> tuple[LongArn, LongArn]:
+    def _generate_map_run_arns(state_machine_arn: Arn, label: str) -> tuple[LongArn, LongArn]:
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/iteration/itemprocessor/map_run_record.py:
```diff
@@ -129,13 +130,13 @@ def __init__(
         # SMA: arn:aws:states:<region>:111111111111:mapRun:<ArnPart_0idx>/<MapRunArnPart0_0idx>
         map_run_arn = state_machine_arn.replace(":stateMachine:", ":mapRun:")
-        part_1 = long_uid()
+        part_1 = long_uid() if label == "" else label
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/label.py:
```diff
@@ -0,0 +1,8 @@
+
+
+class Label(Component):
+    def __init__(self, label: str = ""):
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/label.py:
```diff
@@ -0,0 +1,8 @@
+
+class Label(Component):
+    def __init__(self, label: str = ""):
+        self.label: Final[str] = label
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/state_map.py:
```diff
@@ -80,6 +83,7 @@ class StateMap(ExecutionState):
     catch: Optional[CatchDecl]
     tolerated_failure_count: Optional[ToleratedFailureCount]
     tolerated_failure_percentage: Optional[ToleratedFailurePercentage]
+    label: Label
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/state_map.py:
```diff
@@ -100,6 +104,7 @@ def from_state_props(self, state_props: StateProps) -> None:
         self.result_selector = state_props.get(ResultSelector)
         self.retry = state_props.get(RetryDecl)
         self.catch = state_props.get(CatchDecl)
+        self.label = state_props.get(Label) or Label()
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/state_map.py:
```diff
@@ -189,6 +194,7 @@ def _eval_execution(self, env: Environment) -> None:
                 item_reader=self.item_reader,
                 tolerated_failure_count=tolerated_failure_count,
                 tolerated_failure_percentage=tolerated_failure_percentage,
+                label=self.label.label,
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/state_map.py:
```diff
@@ -212,6 +218,7 @@ def _eval_execution(self, env: Environment) -> None:
                 parameters=self.parameters,
                 tolerated_failure_count=tolerated_failure_count,
                 tolerated_failure_percentage=tolerated_failure_percentage,
+                label=self.label.label,
```
Fixed in [4769662](https://github.com/localstack/localstack/pull/11073/commits/476966293b5118eaca9c167b5319e486b440e5de).

u-minor commented on localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/label.py:
```diff
@@ -0,0 +1,8 @@
+
+class Label(Component):
+    def __init__(self, label: str = ""):
+        self.label: Final[str] = label
```
Fixed in [df3f18d](https://github.com/localstack/localstack/pull/11073/commits/df3f18ddb008ee9572898155f82643f608046ef7).

u-minor commented:
@MEPalma 
I have corrected the areas pointed out in the review and added some validations for label.
I have not checked the actual AWS content for the ValueError error message, should I match it?

MEPalma reviewed (approved):
Thank you for revising the changes and contributing, happy to see Label declarations now being supported!
